Expose metrics via Control

ktoso opened the pull request:
We'd like to be able to expose metrics to users when they use the Sharing the Consumer Actor pattern; I thought this style would likely be good enough - I assume the same could be done in the Publisher?

ktoso commented:
Please review is this general idea could make it in @patriknw @kciesielski @13h3r :)

13h3r commented:
Yes, I think it could be implemented this way

kciesielski commented:
Thanks, LGTM. As for producer, there's `Producer.flow(settings: ProducerSettings, producer: KafkaProducer)`, so users can provide their own producers which can be directly asked for metrics.

ktoso commented:
thanks, will finish this PR Then soon

On Sep 10, 2017 11:47 AM, "Krzysiek Ciesielski" <notifications@github.com>
wrote:

Thanks, LGTM. As for producer, there's Producer.flow(settings:
ProducerSettings, producer: KafkaProducer), so users can provide their own
producers which can be directly asked for metrics.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/akka/reactive-kafka/pull/338#issuecomment-328331691>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAHYk0Vq0nto5M7yUHJOHxkPWNseXxFiks5sg7AfgaJpZM4PPd3A>
.

patriknw commented on core/src/main/scala/akka/kafka/KafkaConsumerActor.scala:
```diff
@@ -48,7 +52,17 @@ object KafkaConsumerActor {
+  /** Public protocol which can be used to interact with the Actor */
+  object Protocol {
+    // requests
+    final case class RequestMetrics()
```
strange with case class without params, case object?

then it need accessor for Java API

ktoso replied:
ok

patriknw commented on core/src/main/scala/akka/kafka/KafkaConsumerActor.scala:
```diff
@@ -48,7 +52,17 @@ object KafkaConsumerActor {
+    
+    // responses
+    /** Contains a snapshot of metrics obtained from the underlying kafka consumer */
+    final case class ConsumerMetrics(metrics: Map[MetricName, Metric])
```
Java API: getMetrics: java.util.Map

v-gerasimov commented:
What about this PR? Seems there is no way to get consumer metrics now.

ktoso commented:
Hah, got reminded of this one via discuss.akka.io Fill fix it

ktoso commented:
Addressed feedback, checked Java API, added docs (also for producer where it's trivial however added it so it's all explained symmetrically).

Please review @kciesielski @patriknw

patriknw commented on core/src/main/scala/akka/kafka/KafkaConsumerActor.scala:
```diff
@@ -61,6 +77,15 @@ object KafkaConsumerActor {
+    override def onPartitionsRevoked(partitions: util.Collection[TopicPartition]): Unit = {
+      onRevoke(partitions.asScala)
+    }
+  }
```
what does this have to do with metrics?

ktoso replied:
Rebase goone weird :\ Thanks

patriknw commented on core/src/main/scala/akka/kafka/KafkaConsumerActor.scala:
```diff
@@ -11,25 +11,41 @@ import java.util.concurrent.TimeUnit
+    /** Java API */
+    def getMetrics: util.Map[MetricName, Metric] = metrics.asJava
+  }
+  // ------------ end of public protocol ------------
```
This works when the KafkaConsumer actor is created and passed in, but what about the other cases where the actor is created by the library? Wouldn't it be better to expose the metrics via the `Control` materialized value?

Could be a Future and using ask internally.

ktoso replied:
Yeah good point, will add that

ktoso commented:
Yeah you're right that a `def metrics` on the `Control` should be added as well; will do so

patriknw commented:
Perhaps `Control` can be the only public API?

ktoso commented:
> Perhaps Control can be the only public API?

Yeah, tried doing that. It's hard to make it `def metrics: Map[MetricName, Metric]` sadly, due to the SubSourceLogic and friends... This would be optimal API for users.

It is easier to make it a `Future[Map` which probably is good enough

ktoso commented:
Done, now the only API is to get metrics from the control value

ktoso commented:
Please give it a look @patriknw when you have some time

chbatey commented on core/src/main/scala/akka/kafka/internal/SingleSourceLogic.scala:
```diff
@@ -6,16 +6,21 @@
-import org.apache.kafka.clients.consumer.ConsumerRecord
-import org.apache.kafka.common.TopicPartition
+import akka.util.Timeout
+import org.apache.kafka.clients.consumer.{Consumer, ConsumerRecord}
```
Consumer is unused

chbatey commented on core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala:
```diff
@@ -9,6 +9,8 @@ import java.util.concurrent.TimeUnit
 
 import akka.NotUsed
 import akka.actor.{ActorRef, ExtendedActorSystem, Terminated}
+import akka.dispatch.ExecutionContexts
```
Clean up imports, some of the new unused

chbatey commented on docs/src/test/scala/sample/scaladsl/ConsumerExample.scala:
```diff
@@ -7,20 +7,23 @@ package sample.scaladsl
 import akka.kafka.ConsumerMessage.{CommittableMessage, CommittableOffsetBatch}
 import akka.kafka._
-import akka.actor.{Props, ActorRef, Actor, ActorSystem, ActorLogging, PoisonPill}
+import akka.actor.{Actor, ActorLogging, ActorRef, ActorSystem, PoisonPill, Props}
```
Some of these new imports are unused

chbatey commented on core/src/main/scala/akka/kafka/KafkaConsumerActor.scala:
```diff
@@ -187,6 +192,11 @@ private[kafka] class KafkaConsumerActor[K, V](settings: ConsumerSettings[K, V])
       }
+
+    case RequestMetrics =>
+      val unmodifiableYetMutableMetrics: java.util.Map[MetricName, _ <: Metric] = consumer.metrics()
```
This map goes through a lot of conversions for the Java case.

I guess there's no point trying to avoid this. Looking at the Kafka code they give us the raw map wrapped in an unmodifiable so we shouldn't pass that around.

ktoso replied:
Yeah it is converted a lot indeed... I figured to leave it along until reported as a problem

ktoso commented:
Thanks! Exciting one :)
